DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+# First class functions
+
+| Section        | Value                                  |
+|----------------|----------------------------------------|
+| DIP:           | 27                                     |
+| Author:        | Amaury SÉCHET                          |
+| Status:        |                                        |
+
+## Abstract
+
+This DIP is the first of 3. It intends to define what is a function in D. The DIP proposes a radically simple approach by deciding that all functions are first class functions. The DIP proposes ways to mitigate breakage involved and allow optional parenthesis when no confusion is possible.
+
+## Description
+
+Function are still defined with the same syntax :
+
+```d
+ReturnType functionName(Parameters) {
+  // Function body.
+}
+```
+
+However, this is now strictly equivalent to :
+
+```d
+enum functionName = function ReturnType(Parameters) {
+  // Function body.
+};
```

Comment:
I thought about that too, but that would make the distinction even fuzzier. Currently `alias` can refer to symbols, templates,  and sequences, wheres `enum` refers to values.
The logical compile-time counterpart of `auto f = (int x) => x + 1` is `enum f = (int x) => x + 1`.
OTOH, `alias fun = x => x + 1` exists, though technically it is a template, not a function pointer value.